.circleci/config.yml

```diff
@@ -0,0 +1,193 @@
+            - v1-source-{{ .Branch }}-
+            - v1-source-
+      - checkout
+      - run: |
+          git submodule sync
+          git submodule update --init
+          chown -R musicbrainz:musicbrainz .
+          sudo -E -H -u musicbrainz git gc
+      - save_cache:
+          key: v1-source-{{ .Branch }}-{{ .Revision }}
+          paths:
+            - ".git"
+
+      - run: |
+          cd postgresql-musicbrainz-collate && make && make install && cd -
+          cd postgresql-musicbrainz-unaccent && make && make install && cd -
+
+      - restore_cache:
+          key: v1-mmd-schema-{{ .Environment.MMD_SCHEMA_TAG }}
+      - run: |
+          cd /home/musicbrainz
+          if [[ ! -d mmd-schema ]]; then
+            git clone https://github.com/metabrainz/mmd-schema.git
+            cd mmd-schema
+          else
+            cd mmd-schema
+            git fetch origin
+          fi
+          git checkout $MMD_SCHEMA_TAG
+          cd ../musicbrainz-server
+      - save_cache:
+          key: v1-mmd-schema-{{ .Environment.MMD_SCHEMA_TAG }}
+          paths:
+            - /home/musicbrainz/mmd-schema
+
+      - restore_cache:
+          keys:
+            - v2-node-{{ checksum "yarn.lock" }}
+            - v2-node-
+      - run: sudo -E -H -u musicbrainz yarn
+      - save_cache:
+          key: v2-node-{{ checksum "yarn.lock" }}
+          paths:
+            - "node_modules"
+
+      - restore_cache:
+          keys:
+            - v3-carton-{{ checksum "cpanfile.snapshot" }}
+            - v3-carton-
+      - run: |
+          sudo -E -H -u musicbrainz carton install --cached --deployment
+          sudo -E -H -u musicbrainz carton bundle
+      - save_cache:
+          key: v3-carton-{{ checksum "cpanfile.snapshot" }}
+          paths:
+            - *carton
+            - /home/musicbrainz/vendor
+
+      - run: |
+          sudo -E -H -u musicbrainz mkdir -p junit_output
+          sudo -E -H -u musicbrainz cp docker/musicbrainz-tests/DBDefs.pm lib/
+          rm /etc/service/{postgresql,redis}/down && sv start postgresql redis
+          sudo -E -H -u musicbrainz carton exec -- ./script/create_test_db.sh
+          sudo -E -H -u musicbrainz make -C po all_quiet deploy
+          sudo -E -H -u musicbrainz carton exec -- ./script/compile_resources.sh default web-tests
+          sudo -E -H -u musicbrainz ./node_modules/.bin/flow --quiet
+          # Prevent postgres from writing to pgdata while
+          # persist_to_workspace runs below.
+          sv stop postgresql
+
+      - persist_to_workspace:
+          root: *home
+          paths:
+            - "carton-local"
+            - "mmd-schema"
+            - "musicbrainz-server"
+            - "pgdata"
+
+workflows:
+  version: 2.1
+  build-and-test:
+    jobs:
+      - js:
+          requires:
+            - build
+      - perl-and-pgtap:
+          requires:
+            - build
+      - selenium:
+          requires:
+            - build
+      - build
```

Comment:
These jobs all run in separate containers, but have a shared "workspace." When the `build` job finishes, it calls `persist_to_workspace` to upload the build result, and other jobs use `attach_workspace` to download a cached copy of it. This is surprisingly quite fast.